### DISCUSSION TITLE
How to parametrise volumes with @script() ?

### DISCUSSION THREAD

**Discussion by EricBoix** [argoproj-labs/hera#738] [Q&A] [answered]
I'm trying to convert some [fully functional in house AW source](https://github.com/VCityTeam/ExpeData-Workflows_testing/blob/master/ArgoWorkflows/Workflows_In_Yaml/CityGMLto3DTiles_Example/workflow-template/atomic-steps.yml#L128) to Hera.
The interesting AW feature boils down to "parametrizing" the `mountPath` of a volume with a template input value, which goes

```bash
  - name: do-some-stuff-template
    inputs:
      parameters:
      - name: persistedVolume
    script:
      image: python:3.9-alpine
      source: |
        <snip ... some python code ...snip>
      volumeMounts:
        - name: workdir
          mountPath: "{{inputs.parameters.persistedVolume}}"          # What is the Hera equivalent ?
```
After taking some inspiration from [Hera's "script auto infer"](https://hera.readthedocs.io/en/stable/examples/workflows/script_auto_infer/) example (and some dark magic incantations, since I have to admit that I the underlying inference mechanism totally eludes me) I dared trying the following:

```python
@script(
    image="python:3.9-alpine",
    volume=Volume(
        name="workdir",
        mount_path=persisted_volume,             # How to transmit this argument (at build time) ?
    ),
)
def do_some_stuff(persisted_volume):          # NameError: name persisted_volume is not defined
     <snip ... some python code ...snip>
```

Alas this fails (not enough chicken blood spilled?). 
The question is thus: how does one pass over parameters to a `@scrip()` configuration section in Hera ?

**Comment by elliotgunton** [comment, marked as answer]
Unfortunately right now the easiest way to do this is a mix of Argo syntax and the `inputs` present in the script decorator (we're working on improving this!)

```py
@script(
    image="python:3.9-alpine",
    inputs=[Parameter(name="persistedVolume")],  # specify an input to the template here instead of as a function arg
    volume=Volume(
        name="workdir",
        mount_path="{{inputs.parameters.persistedVolume}}",  # refer to the input with Argo syntax
    ),
)
def do_some_stuff():          #  persisted_volume (probably) not needed here
     <snip ... some python code ...snip>
```

**Comment by EricBoix** [comment]
Thank you @elliotgunton . Except for the typo (the trailing third curly bracket, `}`, within the `mount_path` definition `mount_path="{{inputs.parameters.persistedVolume}}}"` must be removed) this mix Argo syntax does indeed the trick.
Closing discussion.

**Reply by elliotgunton** [reply]
Oops! Thank you, edited in case people come back to the issue.

**Comment by EricBoix** [comment]
@elliotgunton Also I wanted to mention (at least with version 5.6.0) that it seems that the calling prototype seems to require the mention of the argument. That is one has to write
```python
def do_some_stuff(persisted_volume):          #  persisted_volume MUST be mentioned here 
                                              # (although it is only used in the `@script()` section)
```
Otherwise one gets a `failed to resolve {{inputs.parameters.claim_name}}` at hera submission stage.
But please don't take my words on this since my mileage is really low on this (less than an hour fooling around in a messed up context).

Here is a full example for you to quickly experiment
```python
from hera.workflows import DAG, Parameter, script, Workflow


@script(
    image="{{inputs.parameters.image}}",
)
def test():  # FAILS at submission stage with message:  
             #     failed to resolve {{inputs.parameters.image}}`
             # but the following definitions smoothly accepts submission
             # def test(image):
    ...


with Workflow(generate_name="fail-", entrypoint="main") as w:
    with DAG(name="main"):
        t1 = test(
            arguments=Parameter(
                name="image",
                value="pyton;3.9-alpine",
            ),
        )
w.create()
```

Thanks again.
